client/my-sites/plans-grid/components/comparison-grid/index.tsx

```diff
@@ -805,6 +806,134 @@ const ComparisonGridFeatureGroupRow: React.FunctionComponent< {
 	);
 };
 
+const FeatureGroup = ( {
+	featureGroup,
+	selectedFeature,
+	isInSignup,
+	flowName,
+	intervalType,
+	activeTooltipId,
+	setActiveTooltipId,
+	showUpgradeableStorage,
+	onStorageAddOnClick,
+	featureGroupMap,
+	visibleGridPlans,
+	planFeatureFootnotes,
+}: {
+	featureGroup: FeatureGroup;
+	selectedFeature?: string;
+	isInSignup: boolean;
+	flowName?: string | null;
+	intervalType: string;
+	activeTooltipId: string;
+	setActiveTooltipId: Dispatch< SetStateAction< string > >;
+	showUpgradeableStorage: boolean;
+	onStorageAddOnClick?: ( addOnSlug: WPComStorageAddOnSlug ) => void;
+	featureGroupMap: Partial< FeatureGroupMap >;
+	visibleGridPlans: GridPlan[];
+	planFeatureFootnotes: {
+		footnoteList: string[];
+		footnotesByFeature: Record< string, number >;
+	};
+} ) => {
+	const { allFeaturesList } = usePlansGridContext();
+	const [ firstSetOfFeatures ] = Object.keys( featureGroupMap );
+	const [ visibleFeatureGroups, setVisibleFeatureGroups ] = useState< string[] >( [
+		firstSetOfFeatures,
+	] );
+	const features = featureGroup.get2023PricingGridSignupWpcomFeatures();
+	const featureObjects = getPlanFeaturesObject( allFeaturesList, features );
+	const isHiddenInMobile = ! visibleFeatureGroups.includes( featureGroup.slug );
+
+	const allJetpackFeatures = useMemo( () => {
+		const allPlans = getPlans();
+		const jetpackFeatures = new Set(
+			Object.values( allPlans )
+				.map(
+					( {
+						get2023PricingGridSignupJetpackFeatures,
+						get2023PlanComparisonJetpackFeatureOverride,
+					} ) => {
+						const jetpackFeatures = get2023PricingGridSignupJetpackFeatures?.();
+						const additionalJetpackFeatures = get2023PlanComparisonJetpackFeatureOverride?.();
+
+						return [
+							...( jetpackFeatures ? jetpackFeatures : [] ),
+							...( additionalJetpackFeatures ? additionalJetpackFeatures : [] ),
+						];
+					}
+				)
+				.flat()
+		);
+
+		return jetpackFeatures;
+	}, [] );
+
+	const handleFeatureGroupToggle = useCallback( () => {
```

Comment:
I am not sure. There are at least 20+ passes with same `featureGroup` just on the initial render, not due to interacting with this toggle. So it's hard to tell. The component is subject to re-rendering on any of the props changing, so even `activeTooltipId` changing will force this created anew.

Reply:
> There are at least 20+ passes with same featureGroup just on the initial render,

This has to be because of the parent component re-rendering 20+ times, so I wonder if we are missing the root cause here.

There's a cost to executing `useCallback` as well, along with the cost for react to check if the dependencies have changed.

Looking into this component a bit deeper, I missed something in my first pass - this component does not need to keep the state of all `visibleFeatureGroups`. This component renders a single feature group, so the local state `visibleFeatureGroups` will not be synced anyway between all the `FeatureGroup` components. A single boolean state can replace `visibleFeatureGroups` in this component.

Here's what I mean:
<img width="1888" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/c133ac4b-39ab-4bf9-aa0a-cec187c11afb">

I can put up a PR for this, let me know!

Reply:
> This has to be because of the parent component re-rendering 20+ times, so I wonder if we are missing the root cause here.

As if that's ever not being the case... This component is subject to rendering as many times as its props change, irrespective of what happens in the parent. Even running your mouse pointer over the grid will fire off a few dozen rerenders here.

Reply:
> As if that's ever not been the case... This component is subject to rendering as many times as its props change, irrespective of what happens in the parent.

Apologies for not being clear in my previous comment! What I meant was that we could consider optimizing the rendering of child components when the changes in state or props of the parent component do not affect them. We could explore using `memo` or `children` props for this purpose. I am also curious about the impact of memoizing function callbacks to reduce the render time.`useCallback` may reduce code readability and has some additional cost.

Reply:
I think you would be correct. `TitleRow` is a leaf component, right? Let's do a bit of optimizing here with memo/children on FeatureGroup and remove the memoized callback then. Hopefully that will be better (can we measure "better"?)

> useCallback may reduce code readability

You think?